### PR TITLE
[dry run][dev inspect] Fix dynamic field edge case

### PR DESCRIPTION
**Pull request by tnowacki**
## Description 

Set gas objects in dry run and dev-inspect to the max version to avoid dynamic field issues we have seen with faucet (and faucet tests). 
In dry run and dev inspect, you might load a dynamic field that is actually too new for the transaction. Ideally, we would want to load the "correct" dynamic fields, but as that is not easily determined, we instead can set the sequence number of the gas objects to MAX - 1, to ensure that the lamport version for the transaction is MAX, and as such, a valid lamport version for any object used in the transaction (preventing internal assertions or invariant violations from being triggered)

## Test Plan 

- Added new tests

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

**Comment by tnowacki** [comment]
@patrickkuo @healthydeve, suggestions on fixing the sui-json-rpc tests, e.g. test_public_transfer_object. The transaction digests between dry run and normal execution now differ. I could check that the individual object digests are the same? (except for the gas coins)

**Comment by healthydeve** [comment]
Could we just compare the object changes? If that delta is the same then it should also be fine.

File: crates/sui-core/src/authority.rs [hunk 1]
```diff
@@ -3432,6 +3435,46 @@ fn calculate_checkpoint_numbers(
...
+    input_objects: &mut InputObjects,
+) {
+    const FAKE_SEQ: SequenceNumber = SequenceNumber::from_u64(SequenceNumber::MAX.value() - 1);
+    let mut remapped = HashSet::new();
+    for (id, seq, _digest) in gas_object_refs {
+        remapped.insert(*id);
+        *seq = FAKE_SEQ
+    }
+    let objs = std::mem::replace(input_objects, /* dummy */ InputObjects::new(vec![]));
+    let objs = objs.into_objects();
```
Comment (amnn) [inline crates/sui-core/src/authority.rs:3455]: Seeing as you're going to this trouble, would it be easier to accept `input_objects` by value and return the new one?

File: crates/sui-core/src/authority.rs [hunk 2]
```diff
@@ -3432,6 +3435,46 @@ fn calculate_checkpoint_numbers(
...
+    let fixed_objs = objs
+        .into_iter()
+        .map(|(mut kind, mut object)| {
```
Comment (amnn) [inline crates/sui-core/src/authority.rs:3458]: nit: Potentially slightly more straightforward as `for (kind, object) in &mut fixed_objs { ... }`?

File: crates/sui-json-rpc/src/unit_tests/rpc_server_tests.rs [hunk 3]
```diff
@@ -128,19 +131,40 @@ async fn test_public_transfer_object() -> Result<(), anyhow::Error> {
...
-        effects.unwrap().transaction_digest()
+    assert_same_object_changes_ignoring_version_and_digest(
+        dryrun_response.object_changes,
+        tx_response.object_changes.unwrap(),
     );
-    assert_eq!(dryrun_response.object_changes, object_changes.unwrap());
     Ok(())
 }
 
+fn assert_same_object_changes_ignoring_version_and_digest(
```
Comment (tnowacki) [inline crates/sui-json-rpc/src/unit_tests/rpc_server_tests.rs:141]: @healthydeve @patrickkuo

File: crates/sui-types/src/temporary_store.rs [hunk 4]
```diff
@@ -176,6 +176,36 @@ impl<S> TemporaryStore<S> {
         }
     }
 
+    /// WARNING! Should only be used for dry run and dev inspect!
+    /// In dry run and dev inspect, you might load a dynamic field that is actually too new for
+    /// the transaction. Ideally, we would want to load the "correct" dynamic fields, but as that
+    /// is not easily determined, we instead set the lamport version MAX, which is a valid lamport
+    /// version for any object used in the transaction (preventing internal assertions or
+    /// invariant violations from being triggered)
+    pub fn new_for_mock_transaction(
```
Comment (tnowacki) [inline crates/sui-types/src/temporary_store.rs:185]: I had to change this to not break things like balance changes, so thanks folks for the tests!

File: .github/workflows/e2e.yml [hunk 5]
```diff
@@ -52,14 +52,7 @@ jobs:
 
       - name: Run TS SDK e2e tests
         if: ${{ needs.diff.outputs.isTypescriptSDK == 'true' || needs.diff.outputs.isRust == 'true'}}
-        # usage https://github.com/marketplace/actions/retry-step
```
Comment (tnowacki) [inline .github/workflows/e2e.yml:55]: @666lcz added the revert

File: crates/sui-types/src/temporary_store.rs (hunk 4 above)
Comment (amnn) [reply crates/sui-types/src/temporary_store.rs:185]: This approach seems better, to be honest -- less invasive.

**Pull request merged**